libs/platform/ebpf_ring_buffer.c

```diff
@@ -0,0 +1,261 @@
+_ring_advance_consumer_offset(ebpf_ring_buffer_t* ring, size_t length)
+{
+    ring->consumer_offset += length;
+}
+
+inline static ebpf_ring_buffer_record_t*
+_ring_record_at_offset(ebpf_ring_buffer_t* ring, size_t offset)
+{
+    return (ebpf_ring_buffer_record_t*)&ring->shared_buffer[offset % ring->length];
+}
+
+inline static ebpf_ring_buffer_record_t*
+_ring_next_consumer_record(ebpf_ring_buffer_t* ring)
+{
+    return _ring_record_at_offset(ring, _ring_get_consumer_offset(ring));
+}
+
+inline static ebpf_ring_buffer_record_t*
+_ring_buffer_allocate_record(_In_ ebpf_ring_buffer_t* ring, size_t requested_length)
+{
+    ebpf_ring_buffer_record_t* record = NULL;
+    requested_length += EBPF_OFFSET_OF(ebpf_ring_buffer_record_t, data);
+    size_t remaining_space = ring->length - _ring_get_producer_offset(ring) - _ring_get_consumer_offset(ring);
+
+    if (remaining_space > requested_length) {
+        record = _ring_record_at_offset(ring, _ring_get_producer_offset(ring));
+        _ring_advance_producer_offset(ring, requested_length);
+        record->length = requested_length;
+    }
+    return record;
+}
+
+/**
+ * @brief Free any records in the ring buffer older than ebpf_get_released_epoch as discarded.
+ *
+ * @param[in] context Pointer to the ringbuffer.
+ */
+static void
+_ring_buffer_flush(void* context)
+{
+    ebpf_ring_buffer_t* ring = (ebpf_ring_buffer_t*)context;
+    ebpf_lock_state_t state = ebpf_lock_lock(&ring->lock);
+    while (_ring_get_producer_offset(ring) != _ring_get_consumer_offset(ring)) {
+        ebpf_ring_buffer_record_t* record = _ring_next_consumer_record(ring);
+        // Is this a discarded record?
+        if (record->reserved_epoch == 0) {
+            break;
+            // Is this a reserved record that hasn't expired?
+        } else if (record->reserved_epoch >= ebpf_get_released_epoch()) {
+            break;
+        }
+        _ring_advance_consumer_offset(ring, record->length);
+    }
+    ebpf_lock_unlock(&ring->lock, state);
+}
+
+ebpf_result_t
+ebpf_ring_buffer_create(_Outptr_ ebpf_ring_buffer_t** ring, size_t capacity)
+{
+    ebpf_result_t result;
+    ebpf_ring_buffer_t* local_ring_buffer = ebpf_epoch_allocate(sizeof(ebpf_ring_buffer_t));
+    if (!local_ring_buffer) {
+        result = EBPF_NO_MEMORY;
+        goto Done;
+    }
+
+    if ((capacity & ~(capacity - 1)) != capacity) {
+        result = EBPF_INVALID_ARGUMENT;
+        goto Done;
+    }
+
+    local_ring_buffer->length = capacity;
+
+    local_ring_buffer->ring_descriptor = ebpf_map_ring(capacity);
+    if (!local_ring_buffer->ring_descriptor) {
+        result = EBPF_NO_MEMORY;
+        goto Done;
+    }
+    local_ring_buffer->shared_buffer = ebpf_ring_descriptor_get_base_address(local_ring_buffer->ring_descriptor);
+
+    local_ring_buffer->flush_work_item = ebpf_epoch_allocate_work_item(local_ring_buffer, _ring_buffer_flush);
+
+    *ring = local_ring_buffer;
+    local_ring_buffer = NULL;
+    result = EBPF_SUCCESS;
```

Comment:
nit: would be marginally more efficient to just return EBPF_SUCCESS here (and rename "Done" to "Error").  Up to you.